standard/schema/utils/make_validation_schema.py

```diff
@@ -67,7 +67,7 @@ def add_versions(schema, location=''):
                 version_properties["value"] = new_value
                 schema['properties'][key] = version
 
-        elif prop_type == "object":
+        elif (prop_type == "object" or prop_type == ["object", "null"]) and "properties" in value:
```

Comment:
Can we do a more robust test, where `prop_type` is coerced to a list, and we check for the inclusion of `object` in that list? We can also have `NotImplementedError` for cases/combinations for which we don't have defined behavior.

Reply:
Checking for ```object``` in the list is not good enough due to ```former_value``` which has type ```["string", "number",  "integer", "array",  "object",  "null"]```.
There is also Organization:details that is an `["object",  "null"]` but does not have any properties.
Both these cases probably want to be treated as a whole and is why I chose to be so specific about this case.

I am not sure this tool is the correct place for more in depth schema analysis and make sure we are consistent with our type usage.

Reply:
> Checking for object in the list is not good enough due to…

Indeed - that was my meaning when writing "We can also have `NotImplementedError` for cases/combinations for which we don't have defined behavior." If the defined behavior is to run the default case (for `former_value`, etc.) then the code should obviously continue to do that.

My use case is: "As standard editor, I should not have to keep track of the assumptions various tools are making about the structure of the JSON Schema that I am editing."

As with ocdsmerge, it would be good to better document this code, as it has several specific conditions that would benefit from a comment. In the absence of comments, I'm curious as to whether there should be a special case for an array of strings (which in ocdsmerge is treated as `wholeListMerge` even if `wholeListMerge` is not set https://github.com/open-contracting/ocds-merge/pull/14).

> I am not sure this tool is the correct place for more in depth schema analysis and make sure we are consistent with our type usage.

I don't understand this sentence. Can you expand?

Reply:
What I meant was that we should probably have a tool/test of some kind, separate to these tools, that checks the schema for new or different uses of jsonschema that we currently have not used, or documented, that falls into undefined behaviour. 
I do not think that tool/test should exist in this script or ocdsmerge and they are not necessary the best place to comment on them.

Any new behaviour that falls into default behaviour (without explicit commands) should be documented in http://standard.open-contracting.org/latest/en/schema/merging/

For example if a schema editor writes a `oneOf`, `anyOf` ... then the merge tool, or this tool, will not know what to do with them either.  There are huge amounts of other edge-cases that we probably have not covered.

This will be difficult to write the NotImplementedErrors without guessing all the possible usages a schema editor can write before they write it.

Reply:
I don't think the number of edge cases is that large (the JSON Schema spec is small). If we broke long if-statements into single condition statements, and then added `else: raise NotImplementedError` to each, I think we'd be most of the way there.

I could see some sense in having one thing that identifies new cases by e.g. raising NotImplementError for unexpected cases, instead of making that every tool's responsibility. However, if the number of cases is small, then I might prefer doing it everywhere anyway, as we have no assurance that we'll always remember to run this extra check. Furthermore, even if we have one thing that identifies new cases / lists all cases to implement, that doesn't really help us check that those cases are implemented in all tools. If all tools raised errors for cases they can't handle in the schema, we'd be guaranteed to notice that they aren't handling those cases.

If we are concerned about duplication of effort, we have some options, including:

1. Write a JSON Schema parser with a simple API that emits parts of the schema to tools. If the tool doesn't recognize the part, it raises an error. This centralizes some of the logic and case-handling. It adds an intermediary betweens tools and schema, which has pros and cons.
1. Have all tools call a JSON Schema auditor library during initialization, which – I don't know – returns the 'features' the schema uses, which the tool then compares to its known list of implemented features, and fails if a feature isn't implemented.

There may be other options. However, I wouldn't be confident in a new tool that is entirely divorced from other tools.

Reply:
> I don't think the number of edge cases is that large (the JSON Schema spec is small). If we broke long if-statements into single condition statements, and then added else: raise NotImplementedError to each, I think we'd be most of the way there.

As I said I do think the the amount of edge cases are large.  There 6 different types in jsonschema we have 3 different merge rules and theoretically a schema could have any subset of these.  That is 2^9 = 512 different ways to express what can go into into a field that could effect merging behaviour.  Of course each if/else could cover a large subset of these, but I would not be confident of coving all combinations.  Also there are other cases like not putting in `properties` or `items`, what of the types are that are allowed in a list and constructs like anyOf, oneOf.

When writing the original `make_validation_schema.py` and ocds-merge I wrote a script to analyse all the different combination of types and rules that were actually used and made sure there was logic to cover those in use.  As that seemed like the practical option.

This issue, and the ocds-merge one (https://github.com/open-contracting/ocds-merge/pull/14) have come up due to new combinations of these appearing that had not appeared before those were written.

My point about the test, was to check that schema writers are not adding any new combinations of these things, without considering the consequences to merging behaviour. 

The only way to do that would to be to fix the combinations to the ones we currently use and if there are any new ones then the tests against the standard or an extension should fail.  We could add combinations, but we would have to do that at the same time as modifying the tools.

Only then, with a list of allowed combinations, can the tools be confident that they are not actually doing some undefined behaviour by accident and can raise a `NotImplementedError`. In that case a long list of very specific if/else statements would be workable in all tools.

Reply:
I think having a test makes sense. Can you share the script you wrote to analyze the combination of types and rules, which can perhaps be reused when authoring that test? I imagine we can add the test to https://github.com/open-contracting/standard-maintenance-scripts/blob/master/tests/test_json.py

I started a page documenting what would need to change if we were to change the layout of codelist CSV files, extension.json files, etc. at https://github.com/open-contracting/standard-maintenance-scripts/tree/master/schema I also have a short list of repos that would need to be changed if extensions were changed at http://ocds-standard-development-handbook.readthedocs.io/en/latest/tools/#ocds-extension-support We can consider making a list of things that would need to change if a new pattern is used in the JSON Schema. (I'm also happy to reorganize the above content – I just wrote those to help myself.)